File: rad/engine/02_catalog/store/reclamations.go
package store

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	kv "github.com/Southclaws/rad/rad/engine/01_kv"
	"github.com/Southclaws/rad/rad/engine/01_kv/keyenc"
	"github.com/Southclaws/rad/rad/engine/02_catalog/model"
	"github.com/Southclaws/rad/rad/engine/reject"
)

const (
	reclamationPrefix  = "/rad/catalog/reclamation/"
	reclamationWakeKey = "/rad/catalog/meta/reclamation_seen"
)

func reclamationKey(id string) []byte { return []byte(reclamationPrefix + id) }

// QueueReclamation writes a new durable reclamation record. Re-queuing the
// identical target is idempotent, which lets catalog publication and recovery
// safely reconstruct candidates without creating duplicate physical work.
func QueueReclamation(ctx context.Context, view kv.KV, reclamation model.Reclamation) error {
	if reclamation.State == "" {
		reclamation.State = model.ReclamationPending
	}
	if reclamation.Generation == 0 {
		reclamation.Generation = 1
	}
	if reclamation.CreatedAt.IsZero() {
		reclamation.CreatedAt = time.Now().UTC()
	}
	if reclamation.UpdatedAt.IsZero() {
		reclamation.UpdatedAt = reclamation.CreatedAt
	}
	if err := validateReclamation(reclamation); err != nil {
		return err
	}
	raw, err := json.Marshal(reclamation)
	if err != nil {
		return err
	}
	key := reclamationKey(reclamation.ID)
	existing, ok, err := view.Get(ctx, key)
	if err != nil {
		return err
	}
	if ok {
		current, err := decodeReclamation(reclamation.ID, existing)
		if err != nil {
			return err
		}
		if !current.CompactedAt.IsZero() && current.ID == reclamation.ID && current.Kind == reclamation.Kind {
			return nil
		}
		if sameReclamationTarget(current, reclamation) {
			return nil
		}
		return fmt.Errorf("catalog: reclamation %q already exists with different contents", reclamation.ID)
	}
	if err := view.Put(ctx, key, raw); err != nil {
		return err
	}
	// This sticky marker makes a fresh database's Engine construction a
	// point-read rather than an empty range scan. Once reclamation has ever
	// existed, startup conservatively scans durable records for unfinished
	// work; compacting the tiny history and marker is a later concern.
	return view.Put(ctx, []byte(reclamationWakeKey), []byte{1})
}

func HasReclamationHistory(ctx context.Context, view kv.KV) (bool, error) {
	_, ok, err := view.Get(ctx, []byte(reclamationWakeKey))
	return ok, err
}

func sameReclamationTarget(a, b model.Reclamation) bool {
	if a.ID != b.ID || a.Kind != b.Kind || a.RetiredCatalogVersion != b.RetiredCatalogVersion ||
		a.TableID != b.TableID || a.TableSchemaID != b.TableSchemaID || a.ColumnID != b.ColumnID ||
		a.IndexID != b.IndexID || a.DefinitionGeneration != b.DefinitionGeneration ||
		a.WriteProtocolGeneration != b.WriteProtocolGeneration || a.TransitionID != b.TransitionID ||
		len(a.IndexIDs) != len(b.IndexIDs) {
		return false
	}
	for i := range a.IndexIDs {
		if a.IndexIDs[i] != b.IndexIDs[i] {
			return false
		}
	}
	return true
}

func SaveReclamation(ctx context.Context, view kv.KV, reclamation model.Reclamation) error {
	if err := validateReclamation(reclamation); err != nil {
		return err
	}
	raw, err := json.Marshal(reclamation)
	if err != nil {
		return err
	}
	return view.Put(ctx, reclamationKey(reclamation.ID), raw)
}

func GetReclamation(ctx context.Context, view kv.KV, id string) (model.Reclamation, bool, error) {
	raw, ok, err := view.Get(ctx, reclamationKey(id))
	if err != nil || !ok {
		return model.Reclamation{}, ok, err
	}
	reclamation, err := decodeReclamation(id, raw)
	if err != nil {
		return model.Reclamation{}, false, err
	}
	return reclamation, true, nil
}

func ListReclamations(ctx context.Context, view kv.KV) ([]model.Reclamation, error) {
	prefix := []byte(reclamationPrefix)
	it, err := view.Scan(ctx, prefix, keyenc.PrefixEnd(prefix))
	if err != nil {
		return nil, err
	}
	defer it.Close()
	var reclamations []model.Reclamation
	for it.Next() {
		id := strings.TrimPrefix(string(it.Key()), reclamationPrefix)
		reclamation, err := decodeReclamation(id, it.Value())
		if err != nil {
			return nil, err
		}
		reclamations = append(reclamations, reclamation)
	}
	return reclamations, it.Err()
}

func decodeReclamation(id string, raw []byte) (model.Reclamation, error) {
	reclamation, err := decodeDurableJSON[model.Reclamation]("reclamation", id, raw)
	if err != nil {
		return model.Reclamation{}, err
	}
	if reclamation.ID != id {
		return model.Reclamation{}, reject.Fail(
			reject.ReasonCatalogCorrupt,
			"catalog: reclamation key %q contains ID %q",
			id,
			reclamation.ID,
		)
	}
	if err := validateReclamation(reclamation); err != nil {
		return model.Reclamation{}, reject.Mark(reject.ReasonCatalogCorrupt, err)
	}
	return reclamation, nil
}

func validateReclamation(reclamation model.Reclamation) error {
	if reclamation.ID == "" || reclamation.Kind == "" || reclamation.State == "" || reclamation.Generation == 0 {
		return fmt.Errorf("catalog: invalid reclamation checkpoint %+v", reclamation)
	}
	switch reclamation.State {
	case model.ReclamationPending, model.ReclamationReclaiming, model.ReclamationReclaimed, model.ReclamationFailed:
	default:
		return fmt.Errorf("catalog: reclamation %q has invalid state %q", reclamation.ID, reclamation.State)
	}
	if reclamation.RetiredCatalogVersion == 0 {
		return fmt.Errorf("catalog: reclamation %q has no retirement catalog version", reclamation.ID)
	}
	if reclamation.CreatedAt.IsZero() || reclamation.UpdatedAt.IsZero() {
		return fmt.Errorf("catalog: reclamation %q has invalid checkpoint timestamps", reclamation.ID)
	}
	if !reclamation.CompactedAt.IsZero() && reclamation.State != model.ReclamationReclaimed {
		return fmt.Errorf("catalog: non-reclaimed reclamation %q has a compaction timestamp", reclamation.ID)
	}
	switch reclamation.Kind {
	case model.ReclamationTable:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 {
			return fmt.Errorf("catalog: table reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationColumn:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 || reclamation.ColumnID == "" {
			return fmt.Errorf("catalog: column reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationIndex:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 || reclamation.IndexID == "" {
			return fmt.Errorf("catalog: index reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationTableDefinition:
		if reclamation.TableSchemaID == 0 || reclamation.DefinitionGeneration == 0 {
			return fmt.Errorf("catalog: table-definition reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationWriteProtocolDefinition:
		if reclamation.TableID == "" || reclamation.WriteProtocolGeneration == 0 {
			return fmt.Errorf("catalog: write-protocol reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationTransitionDeltas, model.ReclamationCancelledIndex, model.ReclamationFailedIndex:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 || reclamation.IndexID == "" || reclamation.TransitionID == "" {
			return fmt.Errorf("catalog: transition reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationReplacedColumn, model.ReclamationCancelledReplacement, model.ReclamationFailedReplacement:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 ||
			reclamation.ColumnID == "" || reclamation.TransitionID == "" {
			return fmt.Errorf("catalog: replacement reclamation %q has an incomplete target", reclamation.ID)
		}
	case model.ReclamationConstraintValidation:
		if reclamation.TableID == "" || reclamation.TableSchemaID == 0 || reclamation.TransitionID == "" {
			return fmt.Errorf("catalog: constraint reclamation %q has an incomplete target", reclamation.ID)
		}
	default:
		return fmt.Errorf("catalog: reclamation %q has unknown kind %q", reclamation.ID, reclamation.Kind)
	}
	return nil
}

func TableReclamationID(tableID string) string { return "table-" + tableID }

func ColumnReclamationID(tableID, columnID string) string {
	return "column-" + tableID + "-" + columnID
}

func IndexReclamationID(tableID, indexID string) string {
	return "index-" + tableID + "-" + indexID
}

func TableDefinitionReclamationID(id model.SchemaID, generation uint64) string {
	return fmt.Sprintf("table-definition-%d-%d", id, generation)
}

func WriteProtocolReclamationID(tableID string, generation uint64) string {
	return fmt.Sprintf("write-protocol-%s-%d", tableID, generation)
}

func TransitionDeltaReclamationID(transitionID string) string {
	return "transition-deltas-" + transitionID
}

func CancelledIndexReclamationID(transitionID string) string {
	return "cancelled-index-" + transitionID
}

func FailedIndexReclamationID(transitionID string) string {
	return "failed-index-" + transitionID
}

func ReplacedColumnReclamationID(transitionID string) string {
	return "replaced-column-" + transitionID
}

func CancelledReplacementReclamationID(transitionID string) string {
	return "cancelled-replacement-" + transitionID
}

func FailedReplacementReclamationID(transitionID string) string {
	return "failed-replacement-" + transitionID
}

func ConstraintValidationReclamationID(transitionID string) string {
	return "constraint-validation-" + transitionID
}

File: src/engine/02_catalog/store/reclamations.rs
use bytes::Bytes;

use crate::engine::catalog::identity::{
    ColumnId, DefinitionGeneration, IndexId, ReclamationId, SchemaId, TableId, TransitionId,
    WriteProtocolGeneration,
};
use crate::engine::catalog::model::{Reclamation, ReclamationKind, ReclamationState, Timestamp};
use crate::engine::catalog::{Error, ErrorKind, Result};
use crate::engine::kv::KvView;

use super::durable_json::{decode, encode};
use super::{map_kv, prefix_range};

const RECLAMATION_PREFIX: &str = "/rad/catalog/reclamation/";
const RECLAMATION_WAKE_KEY: &[u8] = b"/rad/catalog/meta/reclamation_seen";

fn reclamation_key(id: &ReclamationId) -> Vec<u8> {
    format!("{RECLAMATION_PREFIX}{id}").into_bytes()
}

pub async fn queue_reclamation<V: KvView + ?Sized>(
    view: &mut V,
    mut reclamation: Reclamation,
    now: Timestamp,
) -> Result<()> {
    if reclamation.generation.is_zero() {
        reclamation.generation = 1.into();
    }
    if reclamation.created_at.is_zero() {
        reclamation.created_at = now;
    }
    if reclamation.updated_at.is_zero() {
        reclamation.updated_at = reclamation.created_at;
    }
    validate_reclamation(&reclamation)?;
    let key = reclamation_key(&reclamation.id);
    if let Some(raw) = view.get(&key).await.map_err(map_kv)? {
        let current = decode_reclamation(reclamation.id.as_str(), &raw)?;
        if (!current.compacted_at.is_zero()
            && current.id == reclamation.id
            && current.kind == reclamation.kind)
            || same_reclamation_target(&current, &reclamation)
        {
            return Ok(());
        }
        return Err(Error::message(
            ErrorKind::Conflict,
            format!(
                "catalog: reclamation {:?} already exists with different contents",
                reclamation.id
            ),
        ));
    }
    let raw = encode("reclamation", reclamation.id.as_str(), &reclamation)?;
    view.put(Bytes::from(key), Bytes::from(raw))
        .await
        .map_err(map_kv)?;
    view.put(
        Bytes::from_static(RECLAMATION_WAKE_KEY),
        Bytes::from_static(&[1]),
    )
    .await
    .map_err(map_kv)
}

pub async fn has_reclamation_history<V: KvView + ?Sized>(view: &mut V) -> Result<bool> {
    Ok(view
        .get(RECLAMATION_WAKE_KEY)
        .await
        .map_err(map_kv)?
        .is_some())
}

pub async fn save_reclamation<V: KvView + ?Sized>(
    view: &mut V,
    reclamation: &Reclamation,
) -> Result<()> {
    validate_reclamation(reclamation)?;
    let raw = encode("reclamation", reclamation.id.as_str(), reclamation)?;
    view.put(
        Bytes::from(reclamation_key(&reclamation.id)),
        Bytes::from(raw),
    )
    .await
    .map_err(map_kv)
}

pub async fn get_reclamation<V: KvView + ?Sized>(
    view: &mut V,
    id: &ReclamationId,
) -> Result<Option<Reclamation>> {
    let Some(raw) = view.get(&reclamation_key(id)).await.map_err(map_kv)? else {
        return Ok(None);
    };
    decode_reclamation(id.as_str(), &raw).map(Some)
}

pub async fn list_reclamations<V: KvView + ?Sized>(view: &mut V) -> Result<Vec<Reclamation>> {
    let prefix = RECLAMATION_PREFIX.as_bytes();
    let mut iterator = view.scan(prefix_range(prefix)).await.map_err(map_kv)?;
    let mut values = Vec::new();
    while let Some(entry) = iterator.next().await.map_err(map_kv)? {
        let id = entry
            .key
            .strip_prefix(prefix)
            .and_then(|value| std::str::from_utf8(value).ok())
            .ok_or_else(|| {
                Error::message(
                    ErrorKind::CatalogCorrupt,
                    "catalog: malformed reclamation key",
                )
            })?;
        values.push(decode_reclamation(id, &entry.value)?);
    }
    Ok(values)
}

fn decode_reclamation(id: &str, raw: &[u8]) -> Result<Reclamation> {
    let reclamation: Reclamation = decode("reclamation", id, raw)?;
    if reclamation.id.as_str() != id {
        return Err(Error::message(
            ErrorKind::CatalogCorrupt,
            format!(
                "catalog: reclamation key {id:?} contains ID {:?}",
                reclamation.id
            ),
        ));
    }
    validate_reclamation(&reclamation)
        .map_err(|error| Error::message(ErrorKind::CatalogCorrupt, error.to_string()))?;
    Ok(reclamation)
}

fn validate_reclamation(value: &Reclamation) -> Result<()> {
    if value.id.is_empty()
        || value.generation.is_zero()
        || value.retired_catalog_version.is_zero()
        || value.created_at.is_zero()
        || value.updated_at.is_zero()
    {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!("catalog: invalid reclamation checkpoint {:?}", value.id),
        ));
    }
    if !value.compacted_at.is_zero() && value.state != ReclamationState::Reclaimed {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!(
                "catalog: non-reclaimed reclamation {:?} has a compaction timestamp",
                value.id
            ),
        ));
    }
    let has_schema_id = value.table_schema_id.is_some();
    let complete = match value.kind {
        ReclamationKind::Table => !value.table_id.is_empty() && has_schema_id,
        ReclamationKind::Column => {
            !value.table_id.is_empty() && has_schema_id && !value.column_id.is_empty()
        }
        ReclamationKind::Index => {
            !value.table_id.is_empty() && has_schema_id && !value.index_id.is_empty()
        }
        ReclamationKind::TableDefinition => has_schema_id && !value.definition_generation.is_zero(),
        ReclamationKind::WriteProtocolDefinition => {
            !value.table_id.is_empty() && !value.write_protocol_generation.is_zero()
        }
        ReclamationKind::TransitionDeltas
        | ReclamationKind::CancelledIndex
        | ReclamationKind::FailedIndex => {
            !value.table_id.is_empty()
                && has_schema_id
                && !value.index_id.is_empty()
                && !value.transition_id.is_empty()
        }
        ReclamationKind::ReplacedColumn
        | ReclamationKind::CancelledReplacement
        | ReclamationKind::FailedReplacement => {
            !value.table_id.is_empty()
                && has_schema_id
                && !value.column_id.is_empty()
                && !value.transition_id.is_empty()
        }
        ReclamationKind::ConstraintValidation => {
            !value.table_id.is_empty() && has_schema_id && !value.transition_id.is_empty()
        }
    };
    if !complete {
        return Err(Error::message(
            ErrorKind::InvalidInput,
            format!(
                "catalog: reclamation {:?} has an incomplete target",
                value.id
            ),
        ));
    }
    Ok(())
}

fn same_reclamation_target(left: &Reclamation, right: &Reclamation) -> bool {
    left.id == right.id
        && left.kind == right.kind
        && left.retired_catalog_version == right.retired_catalog_version
        && left.table_id == right.table_id
        && left.table_schema_id == right.table_schema_id
        && left.column_id == right.column_id
        && left.index_id == right.index_id
        && left.index_ids == right.index_ids
        && left.definition_generation == right.definition_generation
        && left.write_protocol_generation == right.write_protocol_generation
        && left.transition_id == right.transition_id
}

pub fn table_reclamation_id(table_id: &TableId) -> ReclamationId {
    format!("table-{table_id}").into()
}

pub fn column_reclamation_id(table_id: &TableId, column_id: &ColumnId) -> ReclamationId {
    format!("column-{table_id}-{column_id}").into()
}

pub fn index_reclamation_id(table_id: &TableId, index_id: &IndexId) -> ReclamationId {
    format!("index-{table_id}-{index_id}").into()
}

pub fn table_definition_reclamation_id(
    id: SchemaId,
    generation: DefinitionGeneration,
) -> ReclamationId {
    format!("table-definition-{id}-{generation}").into()
}

pub fn write_protocol_reclamation_id(
    table_id: &TableId,
    generation: WriteProtocolGeneration,
) -> ReclamationId {
    format!("write-protocol-{table_id}-{generation}").into()
}

pub fn transition_delta_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("transition-deltas-{id}").into()
}

pub fn cancelled_index_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("cancelled-index-{id}").into()
}

pub fn failed_index_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("failed-index-{id}").into()
}

pub fn replaced_column_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("replaced-column-{id}").into()
}

pub fn cancelled_replacement_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("cancelled-replacement-{id}").into()
}

pub fn failed_replacement_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("failed-replacement-{id}").into()
}

pub fn constraint_validation_reclamation_id(id: &TransitionId) -> ReclamationId {
    format!("constraint-validation-{id}").into()
}

#[cfg(test)]
mod tests {
    use crate::engine::catalog::identity::CatalogVersion;
    use crate::engine::kv::TransactionalKv;
    use crate::engine::kv::slatedb;

    use super::*;

    fn fixture(id: &str) -> Reclamation {
        let mut value = Reclamation::pending(
            ReclamationId::from(id),
            ReclamationKind::TableDefinition,
            CatalogVersion::from(3),
            Timestamp::test_value(),
        );
        value.table_schema_id = Some(SchemaId::new(2).unwrap());
        value.definition_generation = DefinitionGeneration::from(4);
        value
    }

    #[tokio::test]
    async fn queue_is_idempotent_for_the_same_target() {
        let mut database = slatedb::Store::memory("reclamation-idempotence")
            .await
            .unwrap();
        queue_reclamation(&mut database, fixture("r1"), Timestamp::test_value())
            .await
            .unwrap();
        queue_reclamation(&mut database, fixture("r1"), Timestamp::test_value())
            .await
            .unwrap();
        assert!(has_reclamation_history(&mut database).await.unwrap());
        assert_eq!(list_reclamations(&mut database).await.unwrap().len(), 1);
        database.close().await.unwrap();
    }

    #[tokio::test]
    async fn durable_list_rejects_key_identity_drift() {
        let mut database = slatedb::Store::memory("reclamation-key-drift")
            .await
            .unwrap();
        let value = fixture("r1");
        let raw = encode("reclamation", "r1", &value).unwrap();
        database
            .put(
                Bytes::from(reclamation_key(&ReclamationId::from("alias"))),
                Bytes::from(raw),
            )
            .await
            .unwrap();
        assert_eq!(
            list_reclamations(&mut database).await.unwrap_err().kind(),
            ErrorKind::CatalogCorrupt
        );
        database.close().await.unwrap();
    }
}
